### ISSUE TITLE
Reachability analysis of `never` for `yield*` is unchecked

### ISSUE THREAD

**Issue by rixtox** [microsoft/TypeScript#58582]
### 🔎 Search Terms

reachability never 

### 🕗 Version & Regression Information

It's observed in all TS versions.

### ⏯ Playground Link

https://www.typescriptlang.org/play/?#code/GYVwdgxgLglg9mAVAAjAUwG5oE4AoCUAXMgOJrrYCGUc2APOANZhwDuYANOltgHzIBvAFDJkUABbY2yAOTccMgNxCAvkKGhIsBCgC2lGGAKCRY7AE8To0eZhoANgBMU8vPmWiVyYIcr37lsLWyBAIAM5w9mgAdPZwAOa4Mj5gfgEy7qZqoqFgEVGxCUksUMjYaJQQ4pQARlEZympAA

### 💻 Code

```ts
function* never(): Generator<unknown,never> {
  throw 'never';
}

function* main() {
  try {
    yield* never();
  } finally {
    console.log('finally');
  }
  console.log('not reachable');
}
```

### 🙁 Actual behavior

No error reported.

### 🙂 Expected behavior

There should be an error `Unreachable code detected.(7027)` at line 11 `console.log('not reachable');`.

### Additional information about the issue

Reachability analysis is working for normal function: https://www.typescriptlang.org/play/?#code/GYVwdgxgLglg9mABGApgNxQJwBQEoBcy6WiA3gFCKJQAWmcA7ogOSoabMDc5AvueaEiwEiALYBDGGDxlK1TAE9ZVKmyx5uVHomBTxAG31KKKxBAQBnOPpQA6fXADm2ZrrAGjzXJsR8q5sCsbeycXMDgoREwUcQgacQAjGy9uPiA

It's broken for Promise and we have another issue tracking that one: https://github.com/microsoft/TypeScript/issues/34955

Generator functions are being used as roroutine in libraries like [co.js](https://www.npmjs.com/package/co), [Effect.js](https://effect.website/), [redux-saga](https://redux-saga.js.org/), [Effection](https://frontside.com/effection/), and more. It's common to write a coroutine operator that suspend the thread forever, thus taking `function never(): Generator<Instruction, never>` signature. It would benefit users if the TypeScript control analysis can conclude unreachability on `yield* never()`.

**Comment by fatcerberus** [comment]
Related: https://github.com/microsoft/TypeScript/issues/56363

I think this comes down to the known limitation wherein `never`-returning functions can only affect control flow for a naked function call; here, the call is only part of the statement because the return value is being passed to `yield*`.  So it's similar to how

```ts
function never(): never { throw Error(); }
const x = never();
console.log("unreachable, not an error");
```

doesn't have an unreachable-code error either.  Nor does this:

```ts
async function foo(): Promise<never> {
  throw Error();
}

async function bar() {
  await foo();
  console.log("also unreachable, also not an error");
}
```

**Comment by SamVerschueren** [comment]
Run into this as well

```ts
function foo(): never {
  throw new Error('foo');
}

(foo());

const x = foo();

console.log(foo());

foo();

// Only from this point forward it errors because of "unreachable"

console.log('hello');
```

I found this bug while I was looking into [this bug](https://github.com/microsoft/TypeScript/issues/59138) which is somewhat related.
